hello-wiredcraft/src/controllers/user.controller.ts

```diff
@@ -0,0 +1,135 @@
+import {
+  authenticate,
+  TokenService,
+  UserService,
+} from '@loopback/authentication';
+import {inject} from '@loopback/core';
+import {repository} from '@loopback/repository';
+import {
+  del,
+  get,
+  HttpErrors,
+  param,
+  post,
+  put,
+  requestBody,
+} from '@loopback/rest';
+import {SecurityBindings, securityId, UserProfile} from '@loopback/security';
+import _ from 'lodash';
+import {NewUser, User} from '../models';
+import {Credentials, UserRepository} from '../repositories';
+import {TokenServiceBindings, UserServiceBindings} from '../services';
+import {
+  CREATE_USER_REQUEST_SPEC,
+  CREATE_USER_RESPONSE_SPEC,
+  DELETE_USER_RESPONSE_SPEC,
+  GET_USER_ME_RESPONSE_SPEC,
+  GET_USER_RESPONSE_SPEC,
+  UPDATE_USER_REQUEST_SPEC,
+  UPDATE_USER_RESPONSE_SPEC,
+  USER_LOGIN_REQUEST_SPEC,
+  USER_LOGIN_RESPONSE_SPEC,
+} from '../specs/user-spec';
+
+export class UserController {
+  constructor(
+    @repository(UserRepository)
+    public userRepository: UserRepository,
+    @inject(TokenServiceBindings.TOKEN_SERVICE) public jwtService: TokenService,
+    @inject(UserServiceBindings.USER_SERVICE)
+    public userService: UserService<User, Credentials>,
+  ) {}
+
+  @post('/users', CREATE_USER_RESPONSE_SPEC)
+  async create(
+    @requestBody(CREATE_USER_REQUEST_SPEC)
+    newUser: NewUser,
+  ): Promise<User> {
+    const existUser = await this.userRepository.findOne({
+      where: {name: newUser.name},
+    });
+    if (this.isExist(existUser)) {
+      throw new HttpErrors.Conflict('The user is already exists');
+    }
+
+    let user: User = _.omit(newUser, 'password');
+    const password = newUser.password;
+    if (existUser) {
+      user = Object.assign({}, existUser, user, {deleted: false});
+      await this.userRepository.replaceById(user.id, user);
+      await this.userRepository
```

Comment:
Good comment.
I will add a try-catch for it.
When catch error, if the existUser is not null, I will mark the user deleted=true.
Otherwise, I will delete the user and the userCredentials.

Reply:
I don't understand why you do a patch replacement here.
I think you want to replace a deleted user. if you do that why you keep the record in DB but not remove it?

Reply:
Hello Tommy, thanks for your comment.

I add a unique constraint on the user's name and I add a field `deleted` to mark the user soft delete. So if the new user's name exists in the database, I need to do a replacement here.

When deleting a user, I just mark the user deleted but not delete the user credentials yet.
So I need to do a patch replacement here otherwise we will get multiple user credentials when performance user login.

Yes, I can remove the user credentials when deleting the user.
And I just need to create the user credentials here.
But it will end up with a deleted user in DB without a password...